### PR TITLE
Fix MPS SDPA correctness when score matrix exceeds 2^32 elements

### PR DESCRIPTION
**Pull request by jbartolozzi**
## Summary

Fixes #179352. Apple's MPSGraph corrupts SDPA outputs when the score
matrix `B*H*Nq*Nkv` exceeds 2^32 total elements: at the canary shape
(`B=1, H=8, Nq=16384, Nkv=65536, D=64`), output diverges from CPU with
cosine similarity ~0.44 and max relative error ~1.2. The threshold is
independent of dtype, of `is_causal`, and of whether an explicit
`attn_mask` is supplied — the bug is in the matmul/softmax sequence
underneath all four MPSGraph codepaths.

The bug is reachable via `sdpa_general_mps` in
`aten/src/ATen/native/mps/operations/Attention.mm`, which builds an
MPSGraph of the form `matrixMultiplicationWithPrimaryTensor` (Q·Kᵀ) →
optional causal/additive mask → `softMaxWithTensor` →
`matrixMultiplicationWithPrimaryTensor` (·V). The threshold shape
(element-counted, not byte-counted, exactly at 2^32) is consistent
with a 32-bit flat element index somewhere inside Apple's matmul or
softmax kernel, but the bug has not been confirmed against Apple
source. PyTorch's own custom Metal kernels in
`aten/src/ATen/native/mps/kernels/Attention.metal` are not involved
for shapes that hit this bug — `_scaled_dot_product_attention_math_mps`
only dispatches to those when `query_seq_len <= 8`, and the affected
shapes have much larger Q.

Note that `Attention.mm` also defines a third path,
`sdpa_full_attention_mps`, that wires up the custom `attention` Metal
kernel and would not go through MPSGraph at all — but it is dead code
on `main`, never called from the dispatcher (the
`test_fast_full_attention_*` tests are correspondingly all skipped
with "Full attention fast kernel not implemented yet"). Enabling it
was therefore not a viable fix for this PR.

Since SDPA's softmax is row-wise over the Q axis, splitting Q into
chunks and concatenating outputs is mathematically exact.
`sdpa_general_mps` now factors its existing graph build/run into
`sdpa_general_mps_impl` and wraps it: when the score matrix is small
enough (≤ 2^32 elements) it calls the impl once, preserving the
original codepath byte-for-byte; otherwise it loops over Q slices of
size `max(1, 2^32 / (B*H*Nkv))` and stitches the outputs back
together. For `is_causal=True`, the wrapper synthesizes a per-chunk
additive causal mask from the chunk's Q offset and passes it to the
impl with `is_causal=false` (the in-graph causal path assumes the
chunk starts at row 0). User-supplied `attn_mask` is narrowed along
the Q axis. A `TORCH_WARN_ONCE` guards the unrealistic edge case where
`B*H*Nkv` alone exceeds 2^32.

Suggested review order: read the new `sdpa_general_mps_impl` first
(it's the unchanged old body), then the new `sdpa_general_mps`
wrapper, then the test.

## Test plan

- [x] \`python test/test_mps.py -k attention -v\` — 96 tests, all pass
  (skipped=24, expected failures=10, identical to baseline)
- [x] \`python test/test_mps.py -k test_sdpa_large_score_matrix_179352 -v\`
  — 4/4 pass on a 64 GB Mac (3 variants + meta)
- [x] Empirical sweep across dtypes (fp32/fp16/bf16),
  \`is_causal=True/False\`, and bool/additive masks — all formerly
  failing shapes now match CPU to ~1e-5 max relative error
- [x] Lint clean (\`spin lint\`)
- [x] Skipped on CI by design (\`total_memory >= 64 GB\` + \`IS_CI\` gates)

Authored with the assistance of Claude (Anthropic).

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 1]
```diff
@@ -148,6 +150,79 @@
...
+  int64_t kSize = key.size(2);
+  int64_t valueHeadSize = value.size(3);
+
+  auto scale_factor = sdp::calculate_scale(query, scale).expect_float();
+
+  // Apple's MPSGraph corrupts SDPA outputs when the score matrix exceeds 2^32
+  // total elements (pytorch/pytorch#179352). Chunk along the Q axis to keep
+  // each impl call below the threshold. SDPA's softmax is row-wise over Q so
+  // chunking is mathematically exact.
+  constexpr uint64_t kScoreElemThreshold = 1ULL << 32;
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/Attention.mm]: uint32 numeric limit? Encode explicitly as std::numeric_limit?

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 2]
```diff
@@ -148,6 +150,79 @@
...
+      int64_t cur = std::min(q_chunk, qSize - qs);
+      auto q_slice = query.narrow(2, qs, cur);
+
+      std::optional<Tensor> chunk_mask;
+      if (is_causal) {
+        // Synthesize a per-chunk additive causal mask: 0 if j <= (qs + i)
+        // else -1e20 (matching the value used by the in-graph causal path).
+        // Pass it as attn_mask with is_causal=false so the impl's causal
+        // branch (which assumes the chunk starts at row 0) is bypassed.
+        auto i_idx = at::arange(qs, qs + cur, query.options().dtype(kLong)).unsqueeze(1);
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/Attention.mm]: arange can get a bit unstable and miss up indies way before uint64, can't it?

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 3]
```diff
@@ -148,6 +150,79 @@
...
+
+  Tensor out;
+  Tensor attn;
+  if (total_score_elems <= kScoreElemThreshold) {
+    std::tie(out, attn) = sdpa_general_mps_impl(query, key, value, attn_mask, is_causal, scale_factor);
+  } else {
+    uint64_t per_q_row =
+        static_cast<uint64_t>(batchSize) * static_cast<uint64_t>(num_head) * static_cast<uint64_t>(kSize);
+    int64_t q_chunk = static_cast<int64_t>(std::max<uint64_t>(1, kScoreElemThreshold / per_q_row));
+    if (per_q_row > kScoreElemThreshold) {
```
[the remark this answers is not in the document]
Comment (jbartolozzi) [reply aten/src/ATen/native/mps/operations/Attention.mm:192]: Empirically the kernel passes at exactly 2^32 (`B=1, H=8, Nq=8192, Nkv=65536`) and fails at 1.5*2^32, so the original `<=` was intentional. But @Skylion007's comment below pointed out the threshold should be tied to a documented C++ limit, and switching to `std::numeric_limits<uint32_t>::max()` (= 2^32 - 1) addresses both concerns at once: with `T = 2^32 - 1`, both the unchunked branch (`<= T`) and `q_chunk = T / per_q_row` keep the score size strictly below 2^32. Done in the latest fixup commit.

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 4]
```diff
@@ -148,6 +150,79 @@
...
+      std::optional<Tensor> chunk_mask;
+      if (is_causal) {
+        // Synthesize a per-chunk additive causal mask: 0 if j <= (qs + i)
+        // else -1e20 (matching the value used by the in-graph causal path).
+        // Pass it as attn_mask with is_causal=false so the impl's causal
+        // branch (which assumes the chunk starts at row 0) is bypassed.
+        auto i_idx = at::arange(qs, qs + cur, query.options().dtype(kLong)).unsqueeze(1);
+        auto j_idx = at::arange(0, kSize, query.options().dtype(kLong)).unsqueeze(0);
+        auto valid = at::le(j_idx, i_idx);
+        auto m2d = at::where(valid, at::zeros({}, query.options()), at::full({}, -1e20, query.options()));
```
[the remark this answers is not in the document]
Comment (jbartolozzi) [reply aten/src/ATen/native/mps/operations/Attention.mm]: Good catch — the in-graph causal path upcasts `maskedMM` to float32 before applying the additive mask, so its `-1e20` lives in float32, but mine was inheriting `query.options()` and silently becoming `-inf` for fp16 queries. Fixed by building the synthesized mask explicitly in float32. Also took the opportunity to collapse the whole construction to a single `at::triu(at::full({cur, kSize}, -1e20, ...fp32...), qs + 1)` per @Skylion007's comment below — `triu` with `diagonal = qs + 1` keeps positions where `j > qs + i`, exactly the masked-out positions.

File: aten/src/ATen/native/mps/operations/Attention.mm (hunk 1 above)
Comment (jbartolozzi) [reply aten/src/ATen/native/mps/operations/Attention.mm]: Switched to `std::numeric_limits<uint32_t>::max()`. As a side effect this also addresses Copilot's chunk-boundary concern above: with `T = 2^32 - 1` neither the unchunked branch nor any chunked impl call can land at exactly 2^32. The empirical pass-edge case (`B=1, H=8, Nq=8192, Nkv=65536`, total = 2^32) now also gets chunked, but the cost is one extra impl call (q_chunk = 8191) and gives one element of safety margin against the unknown precise kernel limit.

File: aten/src/ATen/native/mps/operations/Attention.mm (hunk 2 above)
Comment (jbartolozzi) [reply aten/src/ATen/native/mps/operations/Attention.mm]: The arange usage is integer (`kLong`) so there's no fp precision issue at our index magnitudes — but your comment made me look harder, and the whole construction collapses to a single `at::triu(at::full({cur, kSize}, -1e20, ...fp32...), qs + 1)`. `triu` with `diagonal = qs + 1` keeps positions where `j - i >= qs + 1`, i.e. `j > qs + i`, which is exactly the masked-out positions; pre-fill with `-1e20` and triu zeroes the kept positions. Drops four ATen op headers and is much easier to read. Replaced in the latest fixup commit.

**Comment by jbartolozzi** [comment]
The `macos-m2-26` test failure is at the `setup-python` action's venv-creation step, not in test code:

> Error: Command `['/Users/ec2-user/runner/_work/_temp/venv-3.12-1777221565/bin/python3.12', '-m', 'ensurepip', '--upgrade', '--default-pip']` returned non-zero exit status 1.

The job exits in 2m15s before any tests run, while `macos-m1-14` and `macos-m2-15` ran the full `test_mps` suite (including the new regression test) and passed in 51-56m. This looks like a Homebrew Python 3.12.13 / ensurepip incompatibility specific to the m2-26 runner image. Could a maintainer re-run that single job, or is this a known infra issue?
